### PR TITLE
Decrease inventory when medication is dispensed

### PR DESCRIPTION
**Pull request by juggernautsei**
Fixes # https://community.open-emr.org/t/drug-inventory-qoh/15955/6

#### Short description of what this resolves:
What this fixes is the tight coupling between the drug dispensing and the drug inventory. The drug inventory can only be changed by selecting the drug dispense button on the prescription page. Immunizations have been left out of this equation.
So, this is an attempt to dispatch an event to update the onhand quantity of the inventory. This way using the event dispatcher will allow the inventory adjustment from anywhere in the system. 

#### Changes proposed in this pull request:
This will add an inventory event to the system...

This is not ready to be included in the codebase. I am posting this PR for peer review purposes. I am still learning and don't have all of the knowledge to create this type of complex system. This will be the bare basics. I just want to know if this will work? 
I need to know a few more things. 

1.  How does the onDispensingItem get called or activated? I am not sure what would make the function execute.
2.  Line 100 of the immunications.php file where the event is supposed to be dispatched and the inventory item sent with the dispatch request. Is that right?
3. In the event subscriber, should the getSubscribedEvents be called in the immunization file?
4. What am I missing or need to add to make this work?

https://symfony.com/doc/current/components/event_dispatcher.html
I followed this tutorial as best I could. There are no clear examples in the system for me to follow that I know of.

### CODE REVIEW

File: interface/patient_file/summary/immunizations.php [hunk 1]
```diff
@@ -93,6 +95,10 @@
         if ($GLOBALS['observation_results_immunization']) {
             saveImmunizationObservationResults($newid, $_GET);
         }
+        if ($GLOBALS['inhouse_pharmacy'] > 0) {
+            $decreaseInventoriedItem = new InventoryDecrementEvent($_GET['lot_number']);
```
Comment (stephenwaite) [inline interface/patient_file/summary/immunizations.php:99]: hi @juggernautsei , have you tried the easy development docker to easily spin up a branch for testing? It should trigger this bit of code and throw a syntax error for missing a `)` at the end of the expression.

`             $decreaseInventoriedItem = new InventoryDecrementEvent($_GET['lot_number']));`

Comment (stephenwaite) [reply interface/patient_file/summary/immunizations.php:99]: wait, it's not missing anything there, sorry

**Comment by sunsetsystems** [comment]
Dunno about the event stuff, but you can't dispense from in-house inventory just by subtracting from QOH. Transactions must be tracked and the operation can fail if there is not enough inventory. There are also issues like choosing the best lot if there are multiple lots available.

To handle these things you must call sellDrug() which is in interface/drugs/drugs.inc.php. See also interface/drugs/dispense_drug.php which is invoked by the prescription module and in turn calls sellDrug().

File: interface/patient_file/summary/immunizations.php (hunk 1 above)
Comment (juggernautsei) [reply interface/patient_file/summary/immunizations.php:99]: Thanks, @stephenwaite. I will fix that syntax error.

**Comment by sunsetsystems** [comment]
By the way after reading the Symfony article that you referenced I'm not seeing how that model contributes to the solution. Perhaps someone can enlighten me.

**Comment by juggernautsei** [comment]
@sunsetsystems thank you for the comment!
You are absolutely right about the inventory change. I have visited the drugs.inc.php file. But to keep this part of the development simplistic at this point. I have elected to not go down that path until the event stuff was in place and working.  I am sure that code can be reused and extended by placing it in a module so that it is broken loose from just the prescriptions.

**Comment by sunsetsystems** [comment]
sellDrug() is also called from the Fee Sheet. Not just for prescriptions.

**Comment by juggernautsei** [comment]
@sunsetsystems I will tell you what I know. 
Right now as it stands the inventory count is only inside the prescription code. As pointed out by Harley here https://community.open-emr.org/t/drug-inventory-qoh/15955/6.  The dispense button is only if the drug is being prescribed through this interface. So, when you look at the immunization which is a dispensing of a drug.  That dispensing does not decrease the inventory on hand. The purpose of the code is to decouple the inventory count from the prescription code. Decoupling to me means instead of using include statements the system dispatches an event. In this case, the immunizations will dispatch an event saying that the inventory needs to be decreased because the drug has been dispensed.  So, instead of adding another include statement in the immunization. The immunization fires off an event. This event is registered by the dispatcher and any module listing to the dispatch can be notified of the event and execute on that event. @sjpadgett uses this in the fax SMS module to change the menu without editing the original code. When the menu is called an event is fired that activates his menu change on the fly. So, what I am attempting to do to create communication within the codebase that allows an inventory event to be registered from anywhere in the system without adding more include statements. The Long-range is that the inventory count will be more flexible and extendable. Then the inventory could be changed from anywhere in the system. That is what I know and trying to accomplish.    @adunsulag is guiding me.

**Comment by juggernautsei** [comment]
https://www.youtube.com/watch?v=WV2Ed1QTst8

**Comment by sunsetsystems** [comment]
Hmm. How are vaccines linking the inventory system at all? I'm not seeing that. The immunization form solicits something it calls a lot number but that seems to be its own thing. And currently the drug add/edit form can link to RXCUI codes, but not CVX codes. Kinda looks like there's a lot to deal with.

**Comment by mdsupport** [comment]
@juggernautsei Suggest staying close to the db design. At the core you have material that is stocked in lots and managed in transactions. So you can implement a material class. Current table is a subclass where material type = drug. This master table has child transactions. Inventory is very similar to billing. Your lot class can provide transaction methods - e.g. $lot->add will bring in vendor, qty, expiry etc and store needed records. $lot->update will add the transaction record. $lot->delete will implement the 'destroy' transaction.
Consumer of the lot object can be shots and fee sheet in current design but it can be several other scenarios. It is misleading to refer to shots as immunization - in a clinic someone may administer B12 shots using inventory which does not have cvx code but does have ndc and other attributes. So the shot screen, apart from doing its own updates, will need $material->list_lots method and will use $lot->update to submit a transaction.
I don't think fee sheet currently handles lots but does update the quantity. So that may need a selection or FIFO approach.

**Comment by juggernautsei** [comment]
As always, thanks @mdsupport. I have no intentions of rewriting what is there. @sunsetsystems  did and an excellent job of creating the inventory module. I just want to extend/modernize the way that it operates in the codebase and elevate my skills.  I truly plan on leaning heavily on @sunsetsystems for merging what I am doing into what is existing. I am still a young pup trying to cut my teeth. As soon as I can get the dispatch system working. I will be calling on @sunsetsystems so we can merge a little bit of code at a time like @stephenwaite  did/does with the billing module.

**Comment by sunsetsystems** [comment]
Correcting my statement above, the Add/Edit Drug page does have a "Relate To" field that can include CVX codes.

It seems reasonable to think that if a practice inventories drugs then it will want to include immunizations with that. The Immunizations page could be modified to notice this, and to support selection of lots from inventory in that case. It would also need to record corresponding entries in the drug_sales table and link to them from the immunizations table.

Also in this case perhaps the Immunizations list can go away, replaced by entries in the drugs table that are linked to a CVX code.

**Comment by juggernautsei** [comment]
I have the code working and will post it later. I am closing this pull request to post a clean code for review.

**Comment by juggernautsei** [comment]
This was a proto type. Now, the working model will be posted.

**Pull request closed without merging**